docs/Guides/Migration-Guide-V5.md

```diff
@@ -18,3 +109,463 @@ behavior and not adhering to [RFC 3986](https://www.rfc-editor.org/rfc/rfc3986#s
+the `parameters` object, such as `toString` or `hasOwnProperty`.
+
+```js
+// v4
+fastify.get('/route', (req, reply) => {
+  console.log(req.params.hasOwnProperty('name'));
+  return { hello: req.params.name };
+});
+```
+
+```js
+// v5
+fastify.get('/route', (req, reply) => {
+  console.log(Object.hasOwn(req.params, 'name'));
+  return { hello: req.params.name };
+});
+```
+
+This increases the security of the application by hardening against prototype
+pollution attacks.
+
+### Type Providers now differentiate between validator and serializer schemas
+
+In v4, the type providers had the same types for both validation and serialization.
+In v5, the type providers have been split into two separate types: `ValidatorSchema`
+and `SerializerSchema`.
+
+[`@fastify/type-provider-json-schema-to-ts`](https://github.com/fastify/fastify-type-provider-json-schema-to-ts)
+and
+[`@fastify/type-provider-typebox`](https://github.com/fastify/fastify-type-provider-typebox)
+have already been updated: upgrade to the latest version to get the new types.
+If you are using a custom type provider, you will need to modify it like
+the following:
+
+```
+--- a/index.ts
++++ b/index.ts
+@@ -11,7 +11,8 @@ import {
+ import { FromSchema, FromSchemaDefaultOptions, FromSchemaOptions, JSONSchema } from 'json-schema-to-ts'
+ 
+ export interface JsonSchemaToTsProvider<
+   Options extends FromSchemaOptions = FromSchemaDefaultOptions
+ > extends FastifyTypeProvider {
+-  output: this['input'] extends JSONSchema ? FromSchema<this['input'], Options> : unknown;
++  validator: this['schema'] extends JSONSchema ? FromSchema<this['schema'], Options> : unknown;
++  serializer: this['schema'] extends JSONSchema ? FromSchema<this['schema'], Options> : unknown;
+ }
+ ```
+
+### Changes to the .listen() method
+
+The variadic argument signature of the `.listen()` method has been removed.
+This means that you can no longer call `.listen()` with a variable number of arguments.
+
+```js
+// v4
+fastify.listen(8000)
+```
+
+Will become:
+
+```js
+// v5
+fastify.listen({ port: 8000 })
+```
+
+This was already deprecated in v4 as `FSTDEP011`, so you should have already updated
+your code to use the new signature.
+
+### Direct return of trailers has been removed
+
+In v4, you could directly return trailers from a handler.
+This is no longer possible in v5.
+
+```js
+// v4
+fastify.get('/route', (req, reply) => {
+  reply.trailer('ETag', function (reply, payload) {
+    return 'custom-etag'
+  })
+  reply.send('')
+});
+```
+
+```js
+// v5
+fastify.get('/route', (req, reply) => {
+  reply.trailer('ETag', async function (reply, payload) {
+    return 'custom-etag'
+  })
+  reply.send('')
+});
+```
+
+A callback could also be used.
+This was already deprecated in v4 as `FSTDEP013`,
+so you should have already updated your code to use the new signature.
+
+### Streamlined access to route definition
+
+Remove all deprecation that related to accessing the route definition
```

Comment:
This line doesn't read well but I can't think of a nicer way to write it.

`All deprecated properties relating to accessing the route definition have been removed`?

Reply:
Maybe we should list the properties being removed?

Reply:
They're being listed in the table below this line.

Reply:
Done